lib/phoenix/verified_routes.ex

```diff
@@ -0,0 +1,719 @@
+
+  defp expand_alias({:__aliases__, _, _} = alias, env),
+    do: Macro.expand(alias, %{env | function: {:path, 2}})
+
+  defp expand_alias(other, _env), do: other
+
+  @doc ~S'''
+  Generates the router path with route verification.
+
+  Interpolated named parameters are encoded via the `Phoenix.Param` protocol.
+
+  Warns when the provided path does not match against the router specified
+  in `use Phoenix.VerifiedRoutes` or the `@router` module attribute.
+
+  ## Examples
+
+      use Phoenix.VerifiedRoutes, endpoint: MyAppWeb.Endpoint, router: MyAppWeb.Router
+
+      redirect(to: ~p"/users/top")
+
+      redirect(to: ~p"/users/#{@user}")
+
+      ~H"""
+      <.link to={~p"/users??page=#{@page}"}>profile</.link>
+
+      <.link to={~p"/users?#{@params}"}>profile</.link>
+      """
+  '''
+  defmacro sigil_p({:<<>>, _meta, _segments} = route, extra) do
+    validate_sigil_p!(extra)
+    endpoint = attr!(__CALLER__, :endpoint)
+    router = attr!(__CALLER__, :router)
+
+    route
+    |> build_route(route, __CALLER__, endpoint, router)
+    |> inject_path(__CALLER__)
+  end
+
+  defp inject_path(
+         {%__MODULE__{} = route, type, _endpoint_ctx, _route_ast, path_ast, static_ast},
+         env
+       ) do
+    case type do
+      :static ->
+        static_ast
+
+      other when other in [:match, :error] ->
+        Module.put_attribute(env.module, :phoenix_verified_routes, route)
+        path_ast
+    end
+  end
+
+  defp inject_url(
+         {%__MODULE__{} = route, type, endpoint_ctx, route_ast, path_ast, _static_ast},
+         env
+       ) do
+    case type do
+      :static ->
+        quote do
+          unquote(__MODULE__).static_url(unquote_splicing([endpoint_ctx, route_ast]))
+        end
+
+      other when other in [:match, :error] ->
+        Module.put_attribute(env.module, :phoenix_verified_routes, route)
+
+        quote do
+          unquote(__MODULE__).unverified_url(unquote_splicing([endpoint_ctx, path_ast]))
+        end
+    end
+  end
+
+  defp validate_sigil_p!([]), do: :ok
+
+  defp validate_sigil_p!(extra) do
+    raise ArgumentError, "~p does not support trailing fragment, got: #{inspect(extra)}"
+  end
+
+  defp raise_invalid_route(ast) do
+    raise ArgumentError,
+          "expected compile-time ~p path string, got: #{Macro.to_string(ast)}\n" <>
+            "Use unverified_path/2 and unverified_url/2 if you need to build an arbitrary path."
+  end
+
+  @doc ~S'''
+  Generates the router path with route verification.
+
+  See `sigil_p/1` for more information.
+
+  Warns when the provided path does not match against the router specified
+  in `use Phoenix.VerifiedRoutes` or the `@router` module attribute.
+
+  ## Examples
+
+      import Phoenix.VerifiedRoutes
+
+      redirect(to: path(conn, MyAppWeb.Router, ~p"/users/top"))
+
+      redirect(to: path(conn, MyAppWeb.Router, ~p"/users/#{@user}"))
+
+      ~H"""
+      <.link to={path(@uri, MyAppWeb.Router, "/users?#{[page: @page]}")}>profile</.link>
```

Comment:
```suggestion
      <.link to={path(@uri, MyAppWeb.Router, "/users?page=#{@page}")}>profile</.link>
```

There may be other uses. :)